src/app/clusters/icd-management-server/ICDManagementCluster.cpp

```diff
@@ -109,141 +88,293 @@ void ICDManagementFabricDelegate::OnFabricRemoved(const FabricTable & fabricTabl
 {
-#if CHIP_CONFIG_ENABLE_ICD_CIP
-    mStorage           = &storage;
-    mSymmetricKeystore = symmetricKeystore;
-    mFabricTable       = &fabricTable;
-#endif // CHIP_CONFIG_ENABLE_ICD_CIP
-    mICDConfigurationData = &icdConfigurationData;
+    MutableCharSpan buffer(mUserActiveModeTriggerInstruction);
+    CopyCharSpanToMutableCharSpanWithTruncation(aUserActiveModeTriggerInstruction, buffer);
 }
 
-CHIP_ERROR ICDManagementAttributeAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+DataModel::ActionReturnStatus ICDManagementCluster::ReadAttribute(const DataModel::ReadAttributeRequest & request,
+                                                                  AttributeValueEncoder & aEncoder)
 {
-    VerifyOrDie(aPath.mClusterId == IcdManagement::Id);
-
-    switch (aPath.mAttributeId)
+    switch (request.path.mAttributeId)
     {
+    case IcdManagement::Attributes::ClusterRevision::Id:
+        return aEncoder.Encode(kRevision);
+
+    case IcdManagement::Attributes::FeatureMap::Id:
+        return aEncoder.Encode(mICDConfigurationData.GetFeatureMap());
+
     case IcdManagement::Attributes::IdleModeDuration::Id:
-        return ReadIdleModeDuration(aPath.mEndpointId, aEncoder);
+        return aEncoder.Encode(mICDConfigurationData.GetIdleModeDuration().count());
 
     case IcdManagement::Attributes::ActiveModeDuration::Id:
-        return ReadActiveModeDuration(aPath.mEndpointId, aEncoder);
+        return aEncoder.Encode(mICDConfigurationData.GetActiveModeDuration().count());
 
     case IcdManagement::Attributes::ActiveModeThreshold::Id:
-        return ReadActiveModeThreshold(aPath.mEndpointId, aEncoder);
+        return aEncoder.Encode(mICDConfigurationData.GetActiveModeThreshold().count());
 
-    case IcdManagement::Attributes::FeatureMap::Id:
-        return ReadFeatureMap(aPath.mEndpointId, aEncoder);
 #if CHIP_CONFIG_ENABLE_ICD_LIT
     case IcdManagement::Attributes::OperatingMode::Id:
-        return ReadOperatingMode(aPath.mEndpointId, aEncoder);
+        return ReadOperatingMode(aEncoder);
 #endif // CHIP_CONFIG_ENABLE_ICD_LIT
-#if CHIP_CONFIG_ENABLE_ICD_CIP
-    case IcdManagement::Attributes::RegisteredClients::Id:
-        return ReadRegisteredClients(aPath.mEndpointId, aEncoder);
 
-    case IcdManagement::Attributes::ICDCounter::Id:
-        return ReadICDCounter(aPath.mEndpointId, aEncoder);
+    case IcdManagement::Attributes::UserActiveModeTriggerHint::Id:
+        return aEncoder.Encode(mUserActiveModeTriggerBitmap);
 
-    case IcdManagement::Attributes::ClientsSupportedPerFabric::Id:
-        return ReadClientsSupportedPerFabric(aPath.mEndpointId, aEncoder);
+    case IcdManagement::Attributes::UserActiveModeTriggerInstruction::Id:
+        return aEncoder.Encode(
+            CharSpan(mUserActiveModeTriggerInstruction,
+                     strnlen(mUserActiveModeTriggerInstruction, IcdManagement::kUserActiveModeTriggerInstructionMaxLength)));
```

Comment:
This right here is a security bug.  Spans are not null-terminated, so this strnlen is going to leak random memory in general.

You need to store the actual length of the data when you copy the data in the constructor.